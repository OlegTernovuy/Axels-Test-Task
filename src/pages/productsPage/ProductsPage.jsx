import React from 'react';
import { Box, Grid } from '@mui/material';

import { GridStyle, StyledDiv } from '../../styled/StyledProductsPage';
import { ProductCard, ProductModal } from '../../components';

const ProductsPage = () => (
  <StyledDiv>
    <Box>
      <ProductModal />
      <Grid container spacing={3} columns={{ xs: 4, sm: 9, md: 12 }}>
        {Array.from(Array(24)).map((_, index) => (
          <GridStyle item xs={2} sm={3} md={3} key={index}>
            <ProductCard />
          </GridStyle>
        ))}
      </Grid>
    </Box>
  </StyledDiv>
);

export default ProductsPage;
