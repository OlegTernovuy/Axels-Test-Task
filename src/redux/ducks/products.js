import { createSlice, createAction } from '@reduxjs/toolkit';
import { put, call, takeEvery } from '@redux-saga/core/effects';

import { getRequest } from '../../api';

const usersInitialState = {
  products: [],
  product: {},
  prodLoading: false,
  singleProdLoading: false,
  errors: '',
};

export const productsSlice = createSlice({
  name: 'products',
  initialState: usersInitialState,
  reducers: {
    getProductsAction: (state, action) => {
      state.errors = '';
      state.prodLoading = true;
    },
    getSingleProductAction: (state, action) => {
      state.errors = '';
      state.singleProdLoading = true;
    },
    getProductSuccessAction: (state, action) => {
      state.products = action.payload;
      state.prodLoading = false;
    },
    getSingleProductSuccessAction: (state, action) => {
      state.product = action.payload;
      state.singleProdLoading = false;
    },
    getProductErrorAction: (state, action) => {
      state.prodLoading = false;
      state.singleProdLoading = false;
      state.errors = action.payload;
    },
  },
});

export const fetchProducts = createAction('products/getProductsAction');

export const fetchSingleProduct = createAction('products/getSingleProductAction');

// Worker
function* getProductsSaga() {
  try {
    const products = yield call(() => getRequest('/prod'));
    yield put(getProductSuccessAction(products));
  } catch (error) {
    yield put(getProductErrorAction(error.message))
  }
}

function* getSingleProductSaga(id) {
  try {
    const product = yield call(() => getRequest(`/prod/${1}`));
    yield put(getSingleProductSuccessAction(product));
  } catch (error) {
    yield put(getProductErrorAction(error.message))
  }
}

// Watcher
export function* watchGetProducts() {
  yield takeEvery(fetchProducts, getProductsSaga);
}

export function* watchGetSingleProduct() {
  yield takeEvery(fetchSingleProduct, getSingleProductSaga);
}

export const {
  getProductsAction,
  getSingleProductAction,
  getProductSuccessAction,
  getSingleProductSuccessAction,
  getProductErrorAction,
} = productsSlice.actions;
export default productsSlice.reducer;
