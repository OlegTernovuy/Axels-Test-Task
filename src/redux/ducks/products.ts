import { createSlice, createAction, PayloadAction } from '@reduxjs/toolkit';
import { put, call, takeEvery } from '@redux-saga/core/effects';
import { type SagaIterator } from '@redux-saga/core';

import { getRequest } from '../../api';
import { IProducts, IProductsState, ISingleProduct } from '../../types';

const usersInitialState: IProductsState = {
  products: [],
  product: {} as ISingleProduct,
  prodLoading: false,
  singleProdLoading: false,
  errors: '',
};

export const products = createSlice({
  name: 'products',
  initialState: usersInitialState,
  reducers: {
    getProductsAction: (state) => {
      state.errors = '';
      state.prodLoading = true;
    },
    getSingleProductAction: (state) => {
      state.errors = '';
      state.singleProdLoading = true;
    },
    getProductSuccessAction: (state, action: PayloadAction<IProducts>) => {
      state.products = action.payload;
      state.prodLoading = false;
    },
    getSingleProductSuccessAction: (state, action: PayloadAction<ISingleProduct>) => {
      state.product = action.payload;
      state.singleProdLoading = false;
    },
    getProductErrorAction: (state, action: PayloadAction<string>) => {
      state.prodLoading = false;
      state.singleProdLoading = false;
      state.errors = action.payload;
    },
  },
});

export const fetchProducts = createAction('Products/getProductsAction');

export const fetchSingleProduct = createAction('Products/getSingleProductAction');

// Worker
function* getProductsSaga(): SagaIterator {
  try {
    const products: IProducts = yield call(() => getRequest<IProducts>('/prod'));    
    yield put(getProductSuccessAction(products));
  } catch (error) {    
    yield put(getProductErrorAction(error.toString()));
  }
}

function* getSingleProductSaga(id): SagaIterator {
  try {
    const product: ISingleProduct = yield call(() => getRequest<ISingleProduct>(`/prod/${1}`));
    yield put(getSingleProductSuccessAction(product));
  } catch (error) {
    yield put(getProductErrorAction(error.toString()))
  }
}

// Watcher
export function* watchGetProducts(): SagaIterator {
  yield takeEvery(fetchProducts, getProductsSaga);
}

export function* watchGetSingleProduct(): SagaIterator {
  yield takeEvery(fetchSingleProduct, getSingleProductSaga);
}

export const {
  getProductsAction,
  getSingleProductAction,
  getProductSuccessAction,
  getSingleProductSuccessAction,
  getProductErrorAction,
} = products.actions;
export default products.reducer;
