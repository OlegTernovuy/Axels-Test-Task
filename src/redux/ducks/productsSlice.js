import { createSlice, createAction } from '@reduxjs/toolkit';
import { put, takeEvery } from '@redux-saga/core/effects';

const usersInitialState = {
  products: [],
  product: {},
  loading: false,
};

export const productsSlice = createSlice({
  name: 'Products',
  initialState: usersInitialState,
  reducers: {
    getProductAction: (state, action) => {
      state.loading = true;
    },
    getProductSuccessAction: (state, action) => {
      state.products = action.payload;
      state.loading = false;
    },
  },
});

export const fetchProducts = createAction('Products/getProductAction');

// Worker
function* getProductSaga() {
  try {
    const response = yield fetch(`http://demo2663713.mockable.io/prod`);
    const json = yield response.json();
    yield put(getProductSuccessAction(json.data));
  } catch (error) {
    console.log(error);
  }
}

// Watcher
export function* watchGetProducts() {
  yield takeEvery(fetchProducts, getProductSaga);
}

export const {
  getProductAction,
  getProductSuccessAction,
  getProductsErrorAction,
} = productsSlice.actions;
export default productsSlice.reducer;
