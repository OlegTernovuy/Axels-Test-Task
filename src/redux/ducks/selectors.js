const products = (store) => store.productsStore.products;
const product = (store) => store.productsStore.product;
const prodLoading = (store) => store.productsStore.prodLoading;
const singleProdLoading = (store) => store.productsStore.singleProdLoading;
const errors = (store) => store.productsStore.errors;

const Selectors = { 
    products,
    product, 
    prodLoading, 
    singleProdLoading, 
    errors
};
export default Selectors;
