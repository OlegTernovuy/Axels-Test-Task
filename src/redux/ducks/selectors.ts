import { RootState } from "..";

const products = (store: RootState) => store.productsStore.products;
const product = (store: RootState) => store.productsStore.product;
const prodLoading = (store: RootState) => store.productsStore.prodLoading;
const singleProdLoading = (store: RootState) => store.productsStore.singleProdLoading;
const errors = (store: RootState) => store.productsStore.errors;

const Selectors = { 
    products,
    product, 
    prodLoading, 
    singleProdLoading, 
    errors
};
export default Selectors;
