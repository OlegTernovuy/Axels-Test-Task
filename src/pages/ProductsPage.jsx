import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Outlet } from 'react-router';
import { Box, Grid } from '@mui/material';

import { GridStyle, StyledDiv } from '../styled/StyledProductsPage';
import { ProductCard } from '../components/index';

import { fetchProducts } from '../redux/ducks/products';
import Selectors from '../redux/ducks/selectors';

const ProductsPage = () => {
  const productsList = useSelector(Selectors.products);
  const loading = useSelector(Selectors.prodLoading);
  const errors = useSelector(Selectors.errors);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchProducts());
  }, [dispatch]);

  return (
    <StyledDiv>
      <Box>
        {loading ? (
          <span>Loading...</span>
        ) : errors.length > 0 ? (
          <>{errors}</>
        ) : (
          <Grid container spacing={3} columns={{ xs: 4, sm: 9, md: 12 }}>
            {productsList &&
              productsList.map((product) => (
                <GridStyle item xs={2} sm={3} md={3} key={product.id}>
                  <ProductCard
                    title={product.title}
                    img={product.img}
                    desc={product.desc}
                    prodId={product.id}
                  />
                </GridStyle>
              ))}
          </Grid>
        )}
      </Box>
      <Outlet />
    </StyledDiv>
  );
};

export default ProductsPage;
