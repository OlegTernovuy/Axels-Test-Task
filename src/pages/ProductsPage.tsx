import { useEffect } from 'react';
import { Outlet } from 'react-router';
import { Box, Grid } from '@mui/material';

import { GridStyle, StyledDiv } from '../styled/StyledProductsPage';
import { ProductCard } from '../components/index';

import { useAppDispatch, useAppSelector } from '../hooks/redux';
import { fetchProducts } from '../redux/ducks/products';
import Selectors from '../redux/ducks/selectors';

const ProductsPage = () => {
  const productsList = useAppSelector(Selectors.products);  
  const loading = useAppSelector(Selectors.prodLoading);
  const errors = useAppSelector(Selectors.errors);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchProducts());
  }, [dispatch]);

  return (
    <StyledDiv>
      <Box>
        {loading ? (
          <span>Loading...</span>
        ) : errors.length > 0 ? (
          <>{errors}</>
        ) : (
          <Grid container spacing={3} columns={{ xs: 4, sm: 9, md: 12 }}>
            {productsList &&
              productsList.map((product) => (
                <GridStyle item xs={2} sm={3} md={3} key={product.id}>
                  <ProductCard
                    title={product.title}
                    img={product.img}
                    desc={product.desc}
                    id={product.id}
                  />
                </GridStyle>
              ))}
          </Grid>
        )}
      </Box>
      <Outlet />
    </StyledDiv>
  );
};

export default ProductsPage;
