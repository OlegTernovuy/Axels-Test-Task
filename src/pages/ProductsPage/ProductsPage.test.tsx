import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import ProductsPage from './ProductsPage';
import { fetchProducts } from '../../redux/ducks/products';

const mockUseDispatch = jest.fn();
const mockUseSelector = jest.fn();

jest.mock('../../redux/ducks/products', () => ({
    fetchProducts: jest.fn(),
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockUseDispatch,
    useSelector: () => mockUseSelector,
}));

describe('Products page snapshot component', () => {
    beforeEach(() => {
        mockUseDispatch.mockClear();
        mockUseSelector.mockClear();
    });
    
    test('Renders loading state correctly', () => {
        render(<ProductsPage />);
        expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
    });

    test('Dispatches fetchProducts on mount', () => {
        render(<ProductsPage />);
        expect(mockUseDispatch).toHaveBeenCalledWith(fetchProducts());
    });

    it('Matches Snapshot', () => {
        const ProductsPageComponent = renderer
            .create(<ProductsPage />)
            .toJSON();
        expect(ProductsPageComponent).toMatchSnapshot();
    });
});
