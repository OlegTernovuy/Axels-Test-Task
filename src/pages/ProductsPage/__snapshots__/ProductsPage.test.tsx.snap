// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products page snapshot component Matches Snapshot 1`] = `
<div
  className="sc-aYaIB ewgwFf"
>
  <div
    className="MuiBox-root css-0"
  >
    <span>
      Loading...
    </span>
  </div>
</div>
`;
